Iteration3/master/app.py

```diff
@@ -0,0 +1,155 @@
+        for ip_address in slaves_ip_addresses():
+            task = asyncio.create_task(
+                replicate_on_slaves(
+                    ip_address, message, done, MESSAGE_ID.value
+                )
+            )
+            logger.info(
+                "Task for replication on slave:%s with ID:%d has been created",
+                ip_address,
+                MESSAGE_ID.value,
+            )
+            common_tasks.append((task, ip_address))
+        MESSAGE_ID.increment()
+
+        await done.wait()
+
+        # For tasks that were not finished we create threads in which
+        # we will make new requests to slaves for replication to ensure that
+        # slave has received our message.
+        if done.write_concern > 1:
+            for task, ip_address in common_tasks:
+                if not task.done():
+                    asyncio.create_task(do_retry_request(ip_address, message))
+    return ",".join(MESSAGES)
+
+
+async def do_retry_request(ip_address, message):
+    retry_cnt = 0
+    while (retry_cnt < MY_RETRY) and (
+        send_target(ip_address, message, MESSAGE_ID.value) != 200
+    ):
+        retry_cnt += 1
+        logger.error(
+            "Retried post to %s | try %d from %d",
+            ip_address,
+            retry_cnt,
+            MY_RETRY,
+        )
+
+
+async def replicate_on_slaves(
+    ip_address: str, message: str, done: asyncio.Event, message_id: int
+):
+    async with httpx.AsyncClient() as client:
+        try:
+            response = await client.post(
+                ip_address,
+                json={"message": message, "message_id": message_id},
+            )
+            logger.info(
+                "Response from the slave:%s with ID:%d has been received",
+                ip_address,
+                message_id,
+            )
+
+            # Basically, the next thing is callback of a coroutine.
+            if response.status_code == 200:
+                done.write_concern -= 1
+            if done.write_concern <= 1:
+                done.set()
+
+            return {
+                "ip_address": f"{ip_address}",
+                "status": response.status_code,
+            }
+        except (httpx.ConnectError, httpx.ReadTimeout) as exception:
+            done.set()
+            logger.error(
+                "Error when making request to the slave:%s with ID:%d %r, "
+                "sent for retry",
+                ip_address,
+                message_id,
+                exception,
+            )
+            return {"ip_address": f"{ip_address}", "status": 500}
+
+
+def send_replication_request(ip_address: str, message: str, message_id: int):
+    """Make replication request in the separate thread."""
+
+    # Now we start target in the separate thread
+    threading.Thread(
+        target=send_target,
+        args=(ip_address, message, message_id),
+        daemon=True,
+    ).start()
+
+
+def send_target(ip_address, message, message_id):
```

Comment:
function name is not the best one as for me, maybe something like `send_message` should be considered. Because current naming is just too confusing.

Reply:
Good point

Reply:
resolved in c78054bc0a202ffb1e889cd723865e480abdcffc by @popenyuk